icefoganalytics/internal-data-portal
# Data Owner: Edit an Existing Dataset

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/7
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- Partially completes https://github.com/icefoganalytics/internal-data-portal/issues/28

# Context

## Dataset Description Manage Page - /dataset/:slug/description/manage

The edit version of the pages doesn’t seem to exist in the wireframes, so I would need another layer of User Role components to handle the different viewer types if I build this out. It might make more sense to just use auto-save for everything.

### Front-end routes

Various tabs on the same level as this component.

### Back-end routes

* /datasets/:id - GET 
* /datasets/:id - PATCH, update data set
* /tags - GET, list tags
* /tags - POST, create new tags
* /taggings - GET, list taggings
* /taggings - POST, associate existing tag with dataset
* /taggings/:id - DELETE, disassociate tag with data, orphaned tags should cleaned up
* /tags - POST, create new tag prior to associating it with a dataset via a tagging
* /users/search?email=...


### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue`
    * Title - Dataset#name
    * `DescriptionTab.vue` - loads DatasetDescriptionEditPage.vue or DatasetDescriptionReadPage.vue
    * `FieldsTab.vue` - loads DatasetFieldsPage.vue
    * `VisualizeTable.vue` - loads DatasetVisualizePage.vue
* `DatasetDescriptionManagePage.vue` - for data owner, sys admin and business analyst
    * `EditButton.vue` - navigates to the edit page?
    * `DataDescriptionFormCard.vue`
        * Name - Dataset#name
        * `AddApiLinkButton.vue` - opens a dialog for updating the Dataset#subscription_url
        * Description - Dataset#description
        * Spatial? - Dataset#is_spacial_data
        * Live data? - Dataset#is_live_data
        * Date Published - Dataset#published_at
        * Inactive checkbox + date field - Dataset#deactivated_at, the checkbox should magically populate the checkbox
        * Terms of Use - Dataset#terms_of_use
        * Credits - Dataset#credits
    * `OwnerFormCard.vue` - may need to diverge from the wireframe slightly to support prefilling with edit after prefilling.
    * `SubjectFormCard.vue`
    * `OwnerNotesFormCard.vue`

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/68e8cbac-e574-4f8d-a048-631d877e8ecc)

# Implementation

Add click through to datasets table.
Build out Data Description card, and dataset saving.
Add a bunch of wiring, and the ability to load/save datasets by slug or id.

# Screenshots

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/e0c2448d-4725-4c46-897b-8fa18a57781c)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Create a new dataset from the /dashboard route.
5. Check that you are redirected to the /description/manage page. e.g. http://localhost:8080/datasets/data-desciption-panel-qa/description/manage
6. Check that the slug matches the dataset name field, but slugified.
7. Check that you can update all the general fields.
8. Check that the "Add API Link" opens a modal with a special api link field, and persists on save.
9. Check that toggline the "inactive" checkbox sets/unsets the "date deactivated" field.
10. Check that the various date picker fields work correctly.
11. Check that you can can go to the "read-only" view of the dataset via the breadcrumbs "Description" link.
12. Check that you can go back to the "manage/edit" view of the dataset via the "Edit" button from the "read-only" view.
13. Check that you can go to the "Datasets" table via the breadcrumbs.
14. Check that each row of the table shows as a link and glows on hover.
15. Check that you still click the table action buttons, and the link feature doesn't break them.
16. Check that you go back to the Dashboard via the breadcrumbs.
